### PR TITLE
Implemented priv_getLogs

### PR DESCRIPTION
**Pull request by lucassaldanha**
#628

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyController.java [hunk 1]
```diff
@@ -236,7 +236,8 @@ private void verifyPrivateFromMatchesEnclavePublicKey(
     }
   }
 
-  private void verifyPrivacyGroupContainsEnclavePublicKey(
+  @Override
+  public void verifyPrivacyGroupContainsEnclavePublicKey(
       final String privacyGroupId, final String enclavePublicKey) {
     final PrivacyGroup privacyGroup = enclave.retrievePrivacyGroup(privacyGroupId);
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyController.java:242]: are you sure that the privacyGroup cannot be null?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateWorldStateReader.java [hunk 2]
```diff
@@ -46,4 +54,20 @@ public PrivateWorldStateReader(
         .flatMap(worldState -> Optional.ofNullable(worldState.get(contractAddress)))
         .flatMap(account -> Optional.ofNullable(account.getCode()));
   }
+
+  public List<PrivateTransactionMetadata> getPrivateTransactionsMetadata(
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateWorldStateReader.java:58]: should that be "getPrivateTransactionMetadataList"?

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyControllerTest.java [hunk 3]
```diff
@@ -418,4 +419,26 @@ public void getContractCodeFailsForInvalidEnclaveKey() {
                     PRIVACY_GROUP_ID, Address.ZERO, Hash.ZERO, ENCLAVE_PUBLIC_KEY2))
         .hasMessage("Privacy group must contain the enclave public key");
   }
+
+  @Test
+  public void verifyPrivacyGroupMatchesEnclaveKeySucceeds() {
+    final PrivacyGroup privacyGroup =
+        new PrivacyGroup(PRIVACY_GROUP_ID, Type.PANTHEON, "", "", List.of(ENCLAVE_PUBLIC_KEY1));
```
Comment (pinges) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyControllerTest.java:426]: Could create a const privacy group ... The same one is used twice.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameter.java [hunk 4]
```diff
@@ -73,6 +73,14 @@ public Hash getBlockhash() {
     return blockhash;
   }
 
+  public boolean isValid() {
+    if (!getFromBlock().isLatest() && !getToBlock().isLatest() && getBlockhash() != null) {
```
[the remark this answers is not in the document]
Comment (lucassaldanha) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/FilterParameter.java:77]: The rules are confusing. Basically, if from/to aren't latest, you can't specify the blockhash.
To be frank, I think we have a bug in this validation. When I compare with the JSON-RPC spec.

The spec says that you can't use fromBlock/toBlock with blockhash at all.

I've talked to Madeline about it and she will raise a JI to review and fix this behaviour.

We are using the exact same logic as eth_getLogs. That's why we decided to keep the same behaviour until we review it. And, if we need to change anything, we change in both methods in the future.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetLogs.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  public JsonRpcResponse response(final JsonRpcRequestContext request) {
+    final String privacyGroupId = request.getRequiredParameter(0, String.class);
+    final FilterParameter filter = request.getRequiredParameter(1, FilterParameter.class);
+
+    checkIfPrivacyGroupMatchesAuthenticatedEnclaveKey(request, privacyGroupId);
+
+    final LogsQuery query =
+        new LogsQuery.Builder().addresses(filter.getAddresses()).topics(filter.getTopics()).build();
+
+    if (!filter.isValid()) {
```
[the remark this answers is not in the document]
Comment (lucassaldanha) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetLogs.java:65]: Sure! :)

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/PrivacyQueries.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+  public List<LogWithMetadata> matchingLogs(
+      final String privacyGroupId, final Hash blockHash, final LogsQuery query) {
+
+    final Optional<BlockHeader> blockHeader = blockchain.getBlockHeader(blockHash);
+    if (blockHeader.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    final List<PrivateTransactionMetadata> privateTransactionMetadatas =
```
[the remark this answers is not in the document]
Comment (lucassaldanha) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/PrivacyQueries.java:67]: Done!

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/PrivacyQueries.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        privateWorldStateReader.getPrivateTransactionsMetadata(privacyGroupId, blockHash);
+
+    final List<PrivateTransactionReceipt> privateTransactionReceipts =
```
[the remark this answers is not in the document]
Comment (lucassaldanha) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/PrivacyQueries.java:70]: Done!

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetLogsTest.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+  @Test
+  public void getMethodReturnsCorrectMethodName() {
+    assertThat(method.getName()).isEqualTo("priv_getLogs");
+  }
+
+  @Test
+  public void privacyGroupIdIsRequired() {
+    final JsonRpcRequestContext request = privGetLogRequest(null, mock(FilterParameter.class));
+
+    final Throwable thrown = catchThrowable(() -> method.response(request));
```
[the remark this answers is not in the document]
Comment (lucassaldanha) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetLogsTest.java:90]: Duh! I always forget about it! Done!

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyController.java (hunk 1 above)
Comment (lucassaldanha) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyController.java:242]: Well, if it is null you won't find any matching privacy group, right?

And in the json-rpc method we are make the privacy group id a required parameter. I don't think we need to be extra defensive here in this internal class.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateWorldStateReader.java (hunk 2 above)
Comment (lucassaldanha) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateWorldStateReader.java:58]: Done

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyControllerTest.java (hunk 3 above)
Comment (lucassaldanha) [reply ethereum/core/src/test/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyControllerTest.java:426]: Done :)

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/contract/CallPrivateSmartContractFunction.java [hunk 9]
```diff
@@ -46,22 +48,65 @@ public CallPrivateSmartContractFunction(
...
   }
 
   @Override
   public String execute(final NodeRequests node) {
     final Besu besu = node.privacy().getBesuClient();
 
-    final PrivateTransactionManager privateTransactionManager =
-        new LegacyPrivateTransactionManager(
-            besu, GAS_PROVIDER, senderCredentials, chainId, privateFrom, privateFor);
+    PrivateTransactionManager privateTransactionManager;
```
Comment (pinges) [inline acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/contract/CallPrivateSmartContractFunction.java:100]: nit: can be final

**Pull request merged**